### DISCUSSION TITLE
Adding task scheduler to a single container

### DISCUSSION THREAD

**Discussion by rburgst** [serversideup/docker-php#297] [Q&A] [answered]
In order to support review apps, I would need to run the scheduler directly within the main php docker container (e.g. most SAAS platforms such as render, vercel or heroku make it very hard to share volumes). As my main laravel container depends on files being generated by the scheduler it would be advantageous (at least for the review app) to run the scheduler in the main container. Since s6 is being used, I would assume this should be possible.

Any guidance would be appreciated.

**Comment by jaydrogers** [comment]
I run two containers:
1. Laravel Web App (responding to HTTPS requests)
2. Laravel Scheduler Container (running the `php artisan schedule:work` command)

My philosophy is one container should do ONE THING very well.

https://serversideup.net/open-source/docker-php/docs/laravel/laravel-task-scheduler

**Reply by jaydrogers** [reply, marked as answer]
I guess my comment above doesn't answer your question fully (sorry, I am rampaging through my inbox now):

> (e.g. most SAAS platforms such as render, vercel or heroku make it very hard to share volumes)

I would highly recommend following my method above.

If you are force to do everything in one container, you'll have to create your own S6 service an monitor it https://github.com/just-containers/s6-overlay

**Reply by tonychuuy** [reply]
> I run two containers:
> 
> 1. Laravel Web App (responding to HTTPS requests)
> 2. Laravel Scheduler Container (running the `php artisan schedule:work` command)
> 
> My philosophy is one container should do ONE THING very well.
> 
> https://serversideup.net/open-source/docker-php/docs/laravel/laravel-task-scheduler

I'm using this method but can't achieve to stop the container gracefully, it waits 10 seconds which is the default, even if no schedule task is running, for example:

```
  task:
    build:
      context: .
      dockerfile: Dockerfile
    volumes:
      - .:/var/www/html
    command: ["php", "artisan", "schedule:work"]
```
I'm using serversideup/php:8.2-fpm, if I have a command like this:

```
Schedule::call(function () {
   echo "test running...";
   sleep(120);
    echo "test finished";
})->everyMinute();
```

Then start the container with `docker compose up -d` it eventually catch the command end execute it, in another terminal I use `docker compose logs -f task`, then run `docker compose stop task`, it will wait the 10 secs and then stop the container no matter if the task is still running. worst if you use the directive `stop_grace_period: 120s` in the docker-compose.yml because it will wait 120 secs for both cases task running/not task running,.

Is there a way to stop the container without interrupt any running tasks? I'm asking this because in order to get the new code (i.e. the new image) I need to stop the container. In a regular Laravel installation you use a cron job if the code is in the file system it will run with the updated code on the next execution. 

Maybe Laravel does not handle SIGTERM for schedule:work like it does for queue:work.

**Reply by jaydrogers** [reply]
I may have published this later, but I have had better results setting the stop_signal to SIGTERM. I updated these examples I think sometime around in September.

If there are better stop signals to use that you find in your tests, let me know https://serversideup.net/open-source/docker-php/docs/laravel/laravel-task-scheduler#examples
